source/agora/consensus/data/Block.d

```diff
@@ -667,3 +699,34 @@ unittest
     assert(merkle_path[3] == habcdefgh);
     assert(block.header.merkle_root == Block.checkMerklePath(hi, merkle_path, 8));
 }
+
+version (unittest)
+{
+    public Block multiSigTestBlock (Block block, KeyPair[] keys, ulong cycle = 0) @trusted nothrow
+    {
+        import agora.common.crypto.Schnorr;
+        import agora.common.crypto.ECC;
+
+        auto validators = BitField!ubyte(keys.length);
+        Signature[] sigs;
+
+        // challenge = Hash(block) to Scalar
+        const Scalar challenge = hashFull(block);
+
+        void validatorSign (ulong i, KeyPair key)
+        {
+            Scalar v = secretKeyToCurveScalar(key.secret);
+            // rc = r used in signing the commitment
+            const Scalar rc = Scalar(hashMulti(v, "consensus.signature.noise", cycle));
```

Comment:
What is `cycle`? Is it supposed to be the current block height? Does it change between calls?

Reply:
I'm guessing it's the index of the validation cycle. first 1008 = 1, second 1008 = 2..

Reply:
I am skeptical about the security of this. Just adding the hash of the block as an offset to R which seems to be static. I think this can be broken, but I don't have a solid test-case for it yet.

Reply:
As long as this is well abstracted away it should be possible to harden the security of this before we deploy CoinNet. So I'm not going to block the PR.

Reply:
Yes I am also not 100% sure if an attack can not be made against this scheme. Would be good to have an audit / brainstorm next year the possible attacks and then harden as required.